geti_sdk/demos/predict_video.py

```diff
@@ -0,0 +1,148 @@
+import cv2
+import ffmpeg
+from tqdm.auto import tqdm
+from tqdm.contrib.logging import logging_redirect_tqdm
+
+from geti_sdk.data_models import Prediction
+from geti_sdk.deployment import Deployment
+from geti_sdk.utils import show_image_with_annotation_scene
+
+
+def predict_video_on_local(
+    video_path: Union[str, os.PathLike],
+    deployment: Union[Deployment, str, os.PathLike],
+    device: str = "CPU",
+    preserve_audio: Optional[bool] = True,
+) -> str:
+    """
+    Create a video reconstruction with overlaid model predictions.
+    This function runs inference on the local machine for every frame in the video.
+    The inference results are overlaid on the frames and the output video path will be returned.
+
+    :param video_path: File path to video
+    :param deployment: Path to the folder containing the Deployment data, or Deployment instance
+    :param device: Device (CPU or GPU) to load the model to. Defaults to 'CPU'
+    :param preserve_audio: True to preserve all audio in the original input video. Defaults to True.
+        If ffmpeg isn't installed, this option is ignored and no audio would be preserved.
+    :return: The file path of the output video if generated successfully. Otherwise None.
+    """
+    retval = None
+
+    # prepare deployment for running inference
+    if isinstance(deployment, (str, os.PathLike)):
+        deployment = Deployment.from_folder(deployment)
+    elif not isinstance(deployment, Deployment):
+        raise ValueError(f"Unable to read deployment {deployment}")
+
+    logging.info("Load inference models")
+    deployment.load_inference_models(device=device)
+
+    # Open the video capture, this prepares the video to be ready for reading
+    cap = cv2.VideoCapture(video_path)
+
+    if cap is None or not cap.isOpened():
+        raise ValueError(f"Unable to read video from {video_path}")
+
+    # Extract original video properties
+    frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+    frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+    fps = cap.get(cv2.CAP_PROP_FPS)
+    num_frames = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
+
+    video_duration = num_frames / fps
+    logging.info(
+        f"Input video contains {num_frames:.1f} frames, "
+        f"for a total duration of {video_duration:.1f} seconds"
+    )
+
+    t_start = time.time()
+
+    predictions: List[Prediction] = []
+    logging.info("Running video prediction... ")
+    with logging_redirect_tqdm(tqdm_class=tqdm), tqdm(
+        total=num_frames, desc="Predicting"
+    ) as progress_bar:
+        while cap.isOpened():
+            ret, frame = cap.read()
+            if ret is True:
+                rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
+                prediction = deployment.infer(rgb_frame)
+                predictions.append((rgb_frame, prediction))
+                progress_bar.update(1)
+            else:
+                break
+
+    cap.release()
+
+    if len(predictions) == num_frames:
```

Comment:
Would be good to log a warning if prediction process failed, i.e. in an `else` clause here